Momentmodels, bindable numerical fluxes, etc

tobiasleibner opened the pull request:
This is #157, transferred to a new branch to see if that fixes CI.

tobiasleibner committed:
start adding momentmodels

tobiasleibner committed:
[test] start fixing mn models

tobiasleibner committed:
[test] fix compilation of mn tests

tobiasleibner committed:
[test] adapt expectations

tobiasleibner committed:
[momentmodels] start cleaning up

tobiasleibner committed:
[entropy] use dynamic jacobian types

tobiasleibner committed:
[momentmodels] improve performance

tobiasleibner committed:
[momentmodels.entropyflux_implementations] use unsafe_add_to_entry

tobiasleibner committed:
[momentmodels] fix tests, start cleaning up

tobiasleibner committed:
[tests.momentmodels.pn] test both reconstruction operators

tobiasleibner committed:
[numerical-fluxes] fix merge conflicts

tobiasleibner committed:
[momentmodels] fix tests

tobiasleibner committed:
[test.pn] fix headercheck

tobiasleibner committed:
[examples] fix compilation of mpi_2019_02_talk_on_hyperbolic_equations

tobiasleibner committed:
[operators.advection] fix segfault in jacobian

tobiasleibner committed:
[tools] fix warnings

tobiasleibner committed:
[test.momentmodels.checkerboard] fix domain definition

tobiasleibner committed:
[operators.reconstruction] fix compilation without Qhull

tobiasleibner committed:
[local.operators] fix automatic jacobians

tobiasleibner committed:
[operators.reconstruction] another compile fix

tobiasleibner committed:
[operators.advection-fv] fix if MPI and reconstruction is used

tobiasleibner committed:
[test] fix passing gridsize as array

tobiasleibner committed:
[operators] add GenericOperator

tobiasleibner committed:
[momentmodels] use dimFlux, rename alpha_iso_prime -> alpha_one

tobiasleibner committed:
[tools.timestepper] fix visualization

tobiasleibner committed:
disambiguate some constructions

tobiasleibner committed:
Merge branch 'master' of github.com:dune-community/dune-gdt into dailywork_tleibner

tobiasleibner committed:
[operators.reconstruction] remove unused include

tobiasleibner committed:
Start adding kinetic scheme

tobiasleibner committed:
fix headercheck

tobiasleibner committed:
[test] Add tests for kinetic scheme

tobiasleibner committed:
[test] Fix tests for kinetic scheme

tobiasleibner committed:
[.ci/shared] bump

tobiasleibner committed:
[momentmodels] start adding different entropy

tobiasleibner committed:
[hyperbolic] make Bose-Einstein entropy work for Mn

tobiasleibner committed:
[hyperbolic] fix partial moment specialization

tobiasleibner committed:
[momentmodels] enable some more tests

tobiasleibner committed:
[momentmodels] fix 1d hatfunction specialization

tobiasleibner committed:
[momentmodels] finish implementing other entropies

tobiasleibner committed:
[momentmodels] some small improvements

tobiasleibner committed:
[hyperbolic] fix compilation of some test

tobiasleibner committed:
[hyperbolic] Fix some errors when dependencies are missing

tobiasleibner committed:
[hyperbolic] increase tolerance in one test

tobiasleibner committed:
[entropyflux] ensure positivity if BoseEinstein entropy is used

tobiasleibner committed:
[entropyflux] silence warnings

tobiasleibner committed:
[momentmodels] improve partial moment basis for large n

tobiasleibner committed:
[momentmodels] Some fixes for high-order moments

tobiasleibner committed:
[tools.adaption-helper] fix compilation for old gcc

tobiasleibner committed:
move momentmodels folder to test

tobiasleibner committed:
[momentmodels] fix some minor errors

tobiasleibner committed:
[momentmodels] remove file

tobiasleibner committed:
[momentmodels] some fixes, new tests

tobiasleibner committed:
[local] fix construction of ProductIntegrand

tobiasleibner committed:
[test] cleanup

renefritze committed:
[ci|gitlab] allow shared runners to pickup builds

renefritze committed:
Revert "[ci|gitlab] allow shared runners to pickup builds"

This reverts commit ac1c1536622a8a7fdb18dfdc93b31353e15d9ba2.

renefritze committed:
[ci] bump

tobiasleibner committed:
[test] add tests for some integrands

tobiasleibner committed:
[test] fix compilation of pn test

tobiasleibner committed:
[integrands] add gradient_value integrand

tobiasleibner committed:
[mapper.continuous] allow for polynomial order 3 or higher

tobiasleibner committed:
[spaces] make spaces copyable for thread safety

tobiasleibner committed:
[entropyflux] fix compilation

tobiasleibner committed:
[test] fix threaded burgers 1d test

tobiasleibner committed:
[local] More fixes for threaded tests

tobiasleibner committed:
[test] cleanup stokes test

tobiasleibner committed:
[test] add test for default_interpolation

tobiasleibner committed:
[finite-elements.wrapper] improve performance

tobiasleibner committed:
[test] fix warnings

tobiasleibner committed:
[local.finite-elements.0d] fix another warning

tobiasleibner committed:
bump ci

tobiasleibner committed:
[integrands] fix quadrature order

tobiasleibner committed:
[discretefunction] allow to enable/disable subsampling

tobiasleibner committed:
Merge branch 'parabolic-lrbms-2018-updates' of github.com:dune-community/dune-gdt into dailywork_tleibner

tobiasleibner committed:
fix headercheck after merge

tobiasleibner committed:
Fix some more merge problems

tobiasleibner committed:
[test] fix elliptic, rename to laplace

tobiasleibner committed:
[test.diffusion-ipdg] fix

tobiasleibner committed:
[test.ESV2007] adapt expectations to corrected quadrature order

tobiasleibner committed:
[entropyflux] adapt to CBLAS changes in xt-common

tobiasleibner committed:
update ci

tobiasleibner committed:
[momentmodels] fix 2d checkerboard

tobiasleibner committed:
clang format

tobiasleibner committed:
[test.momentmodels] fix segfault

tobiasleibner committed:
[examples.adaptive_elliptic_swipdg] fix compilation

tobiasleibner committed:
[integrands.generic] add post_bind function

tobiasleibner committed:
Merge branch 'master' of github.com:dune-community/dune-gdt into momentmodels_etc

tobiasleibner committed:
[operators] fix merge errors

tobiasleibner committed:
[operators.advection] fix copying

tobiasleibner committed:
bump ci

tobiasleibner committed:
[functionals.vector-based] do not turn clang-format off

tobiasleibner commented:
The issues from the last merge should be fixed, so this is again ready to be merged.

ftalbrecht reviewed (changes requested):
I left a bunch of comments. In addition, we should talk about the discrete/nondiscrete operator :/

tobiasleibner commented on dune/gdt/interpolations/default.hh:
```diff
@@ -46,26 +102,18 @@ namespace GDT {
  */
-template <class GV, size_t r, size_t rC, class R, class V, class IGV>
-std::enable_if_t<std::is_same<XT::Grid::extract_entity_t<GV>, typename IGV::Grid::template Codim<0>::Entity>::value,
+template <class GV, size_t r, size_t rC, class R, class V, class IGVT>
```
The ``IGVT`` is the ``Traits`` type, the ``IGV`` the actual ``GridView`` type. Without these changes, there were some errors because the compiler could not match ``Dune::GridView<IGV::Traits>`` and ``IGV``.

tobiasleibner commented on dune/gdt/interpolations/default.hh:
```diff
@@ -84,14 +132,14 @@ void default_interpolation(const XT::Functions::GridFunctionInterface<XT::Grid::
  * \brief Interpolates a localizable function within a given space [creates a suitable target function].
  **/
-template <class VectorType, class GV, size_t r, size_t rC, class R, class IGV>
+template <class VectorType, class GV, size_t r, size_t rC, class R, class IGVT>
```
See above, the ``T`` is for traits.

tobiasleibner commented on dune/gdt/interpolations/default.hh:
```diff
@@ -122,12 +170,12 @@ default_interpolation(const XT::Functions::GridFunctionInterface<XT::Grid::extra
  */
-template <class GV, size_t r, size_t rC, class R, class V, class IGV>
-std::enable_if_t<std::is_same<XT::Grid::extract_entity_t<GV>, typename IGV::Grid::template Codim<0>::Entity>::value,
+template <class GV, size_t r, size_t rC, class R, class V, class IGVT>
```
See above, the ``T`` is for traits.

tobiasleibner commented on dune/gdt/interpolations/default.hh:
```diff
@@ -148,14 +196,14 @@ void default_interpolation(const XT::Functions::FunctionInterface<GV::dimension,
  * \brief Interpolates a function within a given space [creates a suitable target function].
  **/
-template <class VectorType, class GV, size_t r, size_t rC, class R, class IGV>
+template <class VectorType, class GV, size_t r, size_t rC, class R, class IGVT>
```
See above, the ``T`` is for traits.

tobiasleibner commented on dune/gdt/spaces/l2/finite-volume.hh:
```diff
@@ -56,7 +56,7 @@ private:
 
 public:
-  FiniteVolumeSpace(GridViewType grd_vw)
+  FiniteVolumeSpace(GridViewType grd_vw, const int /*order*/ = 0)
```
This was for convenience, I had to create several different spaces somewhere and the other spaces did support this signature. But I agree that it doesn't really make sense, I will remove it again.

tobiasleibner replied:
I added a default order of 1 to the DG space instead.

tobiasleibner commented on dune/gdt/discretefunction/datahandle.hh:
```diff
@@ -0,0 +1,111 @@
+//   Rene Milk      (2018)
+//   Tobias Leibner (2017)
+
+#ifndef DUNE_GDT_DISCRETEFUNCTION_DATAHANDLE_HH
```
Done.

tobiasleibner committed:
[discretefunction] datahandle.hh -> default-datahandle.hh

tobiasleibner commented on dune/gdt/spaces/interface.hh:
```diff
@@ -73,6 +74,11 @@ public:
     , adapted_(false)
   {}
 
+  virtual ThisType* copy() const
```
Done.

tobiasleibner committed:
[spaces] make copy pure virtual

tobiasleibner committed:
remove all virtual from 'virtual ... override'

tobiasleibner commented on dune/gdt/spaces/l2/discontinuous-lagrange.hh:
```diff
@@ -85,7 +85,21 @@ public:
+    this->update_after_adapt();
+  }
+
+  virtual BaseType* copy() const override final
```
I dropped the virtual in all overrides.

tobiasleibner commented on dune/gdt/spaces/hdiv/raviart-thomas.hh:
```diff
@@ -80,7 +80,23 @@ public:
+    this->update_after_adapt();
+  }
+
+  virtual BaseType* copy() const override final
```
I dropped the virtual in all overrides.

tobiasleibner commented on dune/gdt/local/integrands/gradient_value.hh:
```diff
@@ -0,0 +1,176 @@
+//   Tobias Leibner  (2019)
+
+#ifndef DUNE_GDT_LOCAL_INTEGRANDS_QUADRATIC_HH
+#define DUNE_GDT_LOCAL_INTEGRANDS_QUADRATIC_HH
```
Fixed.

tobiasleibner commented on dune/gdt/local/integrands/laplace.hh:
```diff
@@ -68,8 +68,7 @@ public:
   {
-    return local_weight_->order(param) + std::max(test_basis.order(param) - 1, 0)
-           + std::max(ansatz_basis.order(param) - 1, 0);
+    return local_weight_->order(param) + test_basis.order(param) + ansatz_basis.order(param);
```
Yep, actually this makes sense for all our integrands containing derivatives... probably we will never use non-affine geometries anyway, but I copied the warning over.

tobiasleibner commented on dune/gdt/local/integrands/symmetric_elliptic.hh:
```diff
@@ -0,0 +1,140 @@
+//   René Milk       (2017)
+//   Tobias Leibner  (2014, 2016 - 2018)
+
+#ifndef DUNE_GDT_LOCAL_INTEGRANDS_SYMMETRIC_ELLIPTIC_HH
```
I agree, done.

tobiasleibner commented on dune/gdt/local/finite-elements/wrapper.hh:
```diff
@@ -244,7 +244,8 @@ public:
   {
-    return XT::Common::numeric_cast<size_t>(imp_->size());
+    assert(imp_->size() >= 0 && imp_->size() < std::numeric_limits<size_t>::max());
+    return static_cast<size_t>(imp_->size());
```
Unfortunately, the ``numeric_cast`` is quite slow, iirc it was responsible for about 10% of the execution time when I profiled my program. So I replaced the numeric_cast by a static_cast i  functions that are called very often.

tobiasleibner commented on dune/gdt/local/integrands/elliptic.hh:
```diff
@@ -31,7 +31,10 @@ namespace GDT {
 /**
  * Given an inducing scalar function lambda and an inducing matrix-valued function kappa, computes
- * `lambda(x) * {[kappa(x) \nabla phi(x)] * \nabla psi(x)}` for all combinations of phi and psi in the bases.
+ * `lambda(x) * {[kappa(x) \nabla phi(x)] * \nabla psi(x)}` for all combinations of phi in the ansatz basis and psi in
```
I added a note that the class is deprecated and the Laplace integrand should be used instead.

tobiasleibner commented on dune/gdt/local/numerical-fluxes/kinetic.hh:
```diff
@@ -0,0 +1,156 @@
+          direction);
+      for (auto&& entry : ret)
+        if (std::isnan(entry) || std::isinf(entry)) {
+          //          std::cout << XT::Common::to_string(ret) << std::endl;
```
Yep.

tobiasleibner commented on dune/gdt/local/operators/advection-dg.hh:
```diff
@@ -49,26 +49,57 @@ public:
+    , local_flux_(flux_.local_function())
   {}
 
-  /// Applies the inverse of the local mass matrix.
```
Done.

tobiasleibner commented on dune/gdt/local/operators/advection-dg.hh:
```diff
@@ -206,16 +279,17 @@ public:
-    local_flux->bind(intersection.outside());
     numerical_flux_->bind(intersection);
-    const auto outside_flux_order = local_flux->order(param);
-    const auto u = source.local_discrete_function(inside_element);
```
local_source is a member of BaseType, but I did some renaming so it is mostly used as u_ and v_.

tobiasleibner commented on dune/gdt/local/operators/interfaces.hh:
```diff
@@ -78,8 +107,29 @@ public:
+
+protected:
+  XT::Common::ConstStorageProvider<SourceType> source_;
+  std::unique_ptr<LocalSourceType> local_source_;
```
Yes, done.

tobiasleibner commented on dune/gdt/local/operators/advection-fv.hh:
```diff
@@ -48,26 +48,54 @@ template <class I,
+                                   const bool source_x_independent = false)
+    : BaseType(source, numerical_flux.parameter_type())
+    , numerical_flux_(numerical_flux.copy())
+    , source_is_x_independent_(source_x_independent)
```
Constant on each grid element. I renamed it to ``source_is_locally_constant``.

tobiasleibner commented on dune/gdt/local/operators/interfaces.hh:
```diff
@@ -147,11 +225,32 @@ public:
+
+protected:
+  XT::Common::ConstStorageProvider<SourceType> source_;
+  std::unique_ptr<LocalSourceType> local_source_;
```
Done.

tobiasleibner commented on dune/gdt/operators/reconstruction/linear.hh:
```diff
@@ -0,0 +1,488 @@
+//   Rene Milk      (2017 - 2018)
+//   Tobias Leibner (2017)
+
+#ifndef DUNE_GDT_OPERATORS_FV_RECONSTRUCTION_LINEAR_HH
```
Fixed.

tobiasleibner commented on dune/gdt/operators/reconstruction/internal.hh:
```diff
@@ -0,0 +1,521 @@
+//   Rene Milk      (2017 - 2018)
+//   Tobias Leibner (2017)
+
+#ifndef DUNE_GDT_OPERATORS_FV_RECONSTRUCTION_INTERNAL_HH
```
Fixed.

tobiasleibner commented on dune/gdt/operators/reconstruction/linear_kinetic.hh:
```diff
@@ -0,0 +1,182 @@
+// Authors:
+//   Tobias Leibner (2019)
+
+#ifndef DUNE_GDT_OPERATORS_FV_RECONSTRUCTION_LINEAR_KINETIC_HH
```
Fixed.

tobiasleibner commented on dune/gdt/spaces/h1/continuous-lagrange.hh:
```diff
@@ -75,7 +75,21 @@ public:
+    this->update_after_adapt();
+  }
+
+  virtual BaseType* copy() const override final
```
Done.

tobiasleibner commented on dune/gdt/local/assembler/bilinear-form-assemblers.hh:
```diff
@@ -104,17 +104,17 @@ public:
 private:
-  const TestSpaceType& test_space_;
-  const AnsatzSpaceType& ansatz_space_;
+  const std::unique_ptr<TestSpaceType> test_space_;
```
IIRC, the bases are not thread-safe, thus we have to have one space instance in each thread.

tobiasleibner commented on dune/gdt/local/integrands/gradient_value.hh:
```diff
@@ -0,0 +1,176 @@
+  using VectorGridFunctionType = XT::Functions::GridFunctionInterface<E, vector_size, 1, F>;
+  using VectorValues = typename VectorGridFunctionType::LocalFunctionType::RangeReturnType;
+
+  LocalElementGradientValueIntegrand(const XT::Functions::FunctionInterface<d, vector_size, 1, F>& vector_in)
```
Didn't know about that one, nice!

tobiasleibner commented on dune/gdt/operators/advection-fv-with-reconstruction.hh:
```diff
@@ -0,0 +1,142 @@
+//   Rene Milk      (2017 - 2018)
+//   Tobias Leibner (2017)
+
+#ifndef DUNE_GDT_OPERATORS_FV_RECONSTRUCTION_HH
```
Yes.

tobiasleibner commented on dune/gdt/operators/reconstruction/internal.hh:
```diff
@@ -0,0 +1,521 @@
+      for (size_t dd = 0; dd < dimDomain; ++dd) {
+        if (false) {
+          ;
+#if HAVE_MKL || HAVE_LAPACKE
```
There are quite a lot of things that are specific to the flux jacobians in this file. We probably could move parts of it to dune-xt-la, but I do not think it is worth the effort for now.

tobiasleibner commented on dune/gdt/local/operators/interfaces.hh:
```diff
@@ -63,10 +63,39 @@ public:
 
+  using SourceType = XT::Functions::GridFunctionInterface<E, s_r, s_rC, SR>;
+  using LocalSourceType = typename SourceType::LocalFunctionType;
+  using DiscreteSourceType = ConstDiscreteFunction<SV, SGV, s_r, s_rC, SR>;
```
The ``DiscreteLocalSourceType`` was unused, I removed it. The ``DiscreteSourceType`` is only used to be able to construct the local operators with a source space and source vector. I left this unchanged for now, we should do this in #175.

tobiasleibner commented on dune/gdt/local/operators/interfaces.hh:
```diff
@@ -104,18 +154,25 @@ class LocalIntersectionOperatorInterface : public XT::Common::ParametricInterfac
-  using SourceType = ConstDiscreteFunction<SV, SGV, s_r, s_rC, SF>;
+  using SourceType = XT::Functions::GridFunctionInterface<E, s_r, s_rC, SF>;
+  using LocalSourceType = typename SourceType::LocalFunctionType;
+  using DiscreteSourceType = ConstDiscreteFunction<SV, SGV, s_r, s_rC, SF>;
```
see above.

tobiasleibner commented on dune/gdt/operators/interfaces.hh:
```diff
@@ -141,6 +142,7 @@ public:
   static const constexpr size_t r_rC = range_dim_cols;
 
   using SourceSpaceType = SpaceInterface<SGV, s_r, s_rC, F>;
+  using SourceFunctionInterfaceType = XT::Functions::GridFunctionInterface<E, s_r, s_rC, F>;
```
It is used in ``localizable-operator.hh`` in one ``apply`` method. This should also be fixed in #175.

tobiasleibner commented on dune/gdt/operators/localizable-operator.hh:
```diff
@@ -73,10 +73,12 @@ public:
   static const constexpr size_t s_rC = source_range_dim_cols;
   using SF = SourceField;
-  using SourceType = ConstDiscreteFunction<SV, SGV, s_r, s_rC, SF>;
+  using DiscreteSourceType = ConstDiscreteFunction<SV, SGV, s_r, s_rC, SF>;
```
See #175.

tobiasleibner commented on dune/gdt/operators/localizable-operator.hh:
```diff
@@ -302,36 +348,42 @@ public:
   using BaseType::apply;
 
-  void apply(const VectorType& source, VectorType& range, const XT::Common::Parameter& param = {}) const override
+  void apply(const SourceFunctionInterfaceType& source_function,
```
This is #175.

tobiasleibner commented on dune/gdt/local/finite-elements/power.hh:
```diff
@@ -193,10 +193,17 @@ public:
       unpowered_->interpolate(
           [&](const auto& point_in_reference_element) {
-            const auto tmp = local_function(point_in_reference_element);
+            auto cache_it = eval_cache_.find(point_in_reference_element);
```
Yes, that might be a good idea. I opened https://github.com/dune-community/dune-gdt/issues/176 and https://github.com/dune-community/dune-xt/issues/4.

tobiasleibner commented on dune/gdt/operators/generic.hh:
```diff
@@ -0,0 +1,106 @@
+ * \sa OperatorInterface
+ */
+template <class M, class SGV, size_t s_r = 1, size_t s_rC = 1, size_t r_r = s_r, size_t r_rC = s_rC, class RGV = SGV>
+class GenericOperator : public OperatorInterface<M, SGV, s_r, s_rC, r_r, r_rC, RGV>
```
Do you mean the LocalizableDiscreteOperatorApplicator?

tobiasleibner committed:
Address review comments

tobiasleibner committed:
[local.operators] change apply signature of LocalIntersectionOperatorInterface

tobiasleibner committed:
remove template arguments from ThisType to save refactoring work

tobiasleibner committed:
[local.operator] adapt to new apply signature

tobiasleibner committed:
[local] add some missing binds

tobiasleibner committed:
[local.operators] do not use intersection.outside() if intersection.neighbor() is false

tobiasleibner committed:
[operators.advection-fv-entropybased] fix

tobiasleibner commented:
@ftalbrecht, I addressed your comments and there are some new open issues now for the parts we decided to do after the merge. Please have a look at the new commits and feel free to merge this PR if you are okay with these changes.

ftalbrecht commented on dune/gdt/local/integrands/generic.hh:
```diff
@@ -33,41 +33,50 @@ public:
-                const DomainType& point_in_reference_element,
-                DynamicVector<F>& result,
-                const XT::Common::Parameter& param = {}) const
+  virtual void evaluate(const LocalBasisType& basis,
```
The `virtual` seems to be still there

ftalbrecht reviewed (approved):
A few virtuals seem to be still there, apart from that I am happy!

tobiasleibner commented:
Yep, my search and replace did not take into account functions where ``virtual`` and ``override`` are on different lines. I will remove the remaining ``virtual``s manually whenever I see them.
